### ISSUE TITLE
dsolve should solve different factors independently

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#10043]
For example, `eq = Eq(-M * phi(t).diff(t), Rational(3, 2) * m * r**2 * phi(t).diff(t) * phi(t).diff(t,t))` (from https://stackoverflow.com/questions/33340217/sympy-solve-a-differential-equation). You can factor out a `phi(t).diff(t)` and solve the two factors independently.

One question is how to represent the solutions. `dsolve` already returns a list when its internal use of `solve` does, so that is probably fine. But can the two solutions both use the same C1 constant, or should there be three constants?

**Comment by sudeepkatakol** [comment]
@asmeurer  
"dsolve already returns a list when its internal use of solve does"
When does this case occur?
Because
```
In [33]: f=(y(x).diff(x) -3)*(y(x).diff(x)-3*x)

In [34]: f
Out[34]: 
⎛       d       ⎞ ⎛d           ⎞
⎜-3⋅x + ──(y(x))⎟⋅⎜──(y(x)) - 3⎟
⎝       dx      ⎠ ⎝dx          ⎠

In [35]: dsolve(f)
Out[35]: y(x) = C₁ + 3⋅x
```

**Comment by asmeurer** [comment]
That happens when dsolve solves for f(x). For example:

```
In [66]: dsolve(2*diff(f(x), x)*x*f(x) + f(x)**2 + 1)
Out[66]:
⎡           ________              ________⎤
⎢          ╱ C₁                  ╱ C₁     ⎥
⎢f(x) =   ╱  ── - 1 , f(x) = -  ╱  ── - 1 ⎥
⎣       ╲╱   x                ╲╱   x      ⎦
```

**Comment by oscarbenjamin** [comment]
The equation above now matches the `nth_algebraic` solver which is able to solve it but gets confused in the code for handling redundant solutions:

```julia
In [1]: M, r = symbols('M r')

In [2]: phi = Function('phi')

In [3]: eq = Eq(-M * phi(t).diff(t), Rational(3, 2) * m * r**2 * phi(t).diff(t) * phi(t).diff(t,t))

In [4]: eq
Out[4]: 
                                2      
                   2 d         d       
              3⋅m⋅r ⋅──(φ(t))⋅───(φ(t))
                     dt         2      
   d                          dt       
-M⋅──(φ(t)) = ─────────────────────────
   dt                     2            

In [5]: dsolve(eq)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-5-42dc4d2e7b25> in <module>()
----> 1 dsolve(eq)

~/current/sympy/sympy/sympy/solvers/ode.py in dsolve(eq, func, hint, simplify, ics, xi, eta, x0, n, **kwargs)
    674             # The key 'hint' stores the hint needed to be solved for.
    675             hint = hints['hint']
--> 676             return _helper_simplify(eq, hint, hints, simplify, ics=ics)
    677 
    678 def _helper_simplify(eq, hint, match, simplify=True, ics=None, **kwargs):

~/current/sympy/sympy/sympy/solvers/ode.py in _helper_simplify(eq, hint, match, simplify, ics, **kwargs)
    699         # attempt to solve for func, and apply any other hint specific
    700         # simplifications
--> 701         sols = solvefunc(eq, func, order, match)
    702         if isinstance(sols, Expr):
    703             rv =  odesimp(sols, func, order, cons(sols), hint)

~/current/sympy/sympy/sympy/solvers/ode.py in ode_nth_algebraic(eq, func, order, match)
   4068     solns = match['solutions']
   4069     var = match['var']
-> 4070     solns = _nth_algebraic_remove_redundant_solutions(eq, solns, order, var)
   4071     if len(solns) == 1:
   4072         return solns[0]

~/current/sympy/sympy/sympy/solvers/ode.py in _nth_algebraic_remove_redundant_solutions(eq, solns, order, var)
   4118             solns_integral[soln] = constants
   4119         else:
-> 4120             assert False, "Solution should have 0 or order constants..."
   4121 
   4122     # Compare each algebraic solution with each integral solution to remove

AssertionError: Solution should have 0 or order constants...
```

From pdb the solutions that would be returned are
```julia
In [15]: s1.doit()
Out[15]: φ(t) = C₂

In [16]: s2.doit()
Out[16]: 
                       2 
                    M⋅t  
φ(t) = C₁⋅t + C₂ - ──────
                        2
                   3⋅m⋅r 
```
which looks correct to me.

**Comment by oscarbenjamin** [comment]
The issue as described is actually not fixed even if the example given above is now solved by `nth_algebraic`. Here is another example that should be solved as independent factors (giving two solutions):

```julia
In [5]: eq = (f(x).diff(x)-f(x)) * (f(x).diff(x)+f(x))

In [6]: eq
Out[6]: 
⎛        d       ⎞ ⎛       d       ⎞
⎜-f(x) + ──(f(x))⎟⋅⎜f(x) + ──(f(x))⎟
⎝        dx      ⎠ ⎝       dx      ⎠

In [7]: dsolve(eq, f(x))
Out[7]: 
           -x
f(x) = C₁⋅ℯ  

```

**Comment by oscarbenjamin** [comment]
Another example:
```julia
In [4]: eq = f(x).diff(x) * (f(x).diff(x) - f(x))                                                                                              

In [5]: dsolve(eq, f(x))                                                                                                                       
Out[5]: f(x) = C₁

In [6]: classify_ode(eq)                                                                                                                       
Out[6]: ('lie_group',)
```

**Comment by oscarbenjamin** [comment]
And another:
```julia
In [18]: eq = f(x).diff(x)**2 - f(x)**2                                                                                                        

In [19]: dsolve(eq, f(x))                                                                                                                      
Out[19]: 
           -x
f(x) = C₁⋅ℯ    # Missing a solution

In [20]: classify_ode(eq)                                                                                                                      
Out[20]: ('lie_group',)
```

This example factors as
```julia
In [21]: factor(eq)                                                                                                                            
Out[21]: 
⎛        d       ⎞ ⎛       d       ⎞
⎜-f(x) + ──(f(x))⎟⋅⎜f(x) + ──(f(x))⎟
⎝        dx      ⎠ ⎝       dx      ⎠
```

**Comment by RituRajSingh878** [comment]
I am trying to solve this issue and I am proposing the solution for this issue is, after getting differential equation we should factorize the equation and if those factors are differential equation we will pass them one by one for solving.
e.g. - 
```
In [61]: eq = f(x).diff(x)**2 - f(x)**2                                         

In [62]: eq                                                                     
Out[62]: 
                    2
   2      ⎛d       ⎞ 
- f (x) + ⎜──(f(x))⎟ 
          ⎝dx      ⎠ 

In [63]: factor(eq)                                                             
Out[63]: 
⎛        d       ⎞ ⎛       d       ⎞
⎜-f(x) + ──(f(x))⎟⋅⎜f(x) + ──(f(x))⎟
⎝        dx      ⎠ ⎝       dx      ⎠

```

here will pass factor1 and factor2.

eq 2-

```
In [64]: eq = x*f(x).diff(x) - x**2*f(x)                                        

In [65]: eq                                                                     
Out[65]: 
   2          d       
- x ⋅f(x) + x⋅──(f(x))
              dx      

In [66]: factor(eq)                                                             
Out[66]: 
   ⎛         d       ⎞
-x⋅⎜x⋅f(x) - ──(f(x))⎟
   ⎝         dx      ⎠


```

in this example we will pass only second factor for solving;


So I think it will solve most of the problem and it will remove unnecessary power of x also.

**Comment by RituRajSingh878** [comment]
All of the above problems told by @oscarbenjamin is due to the way `lie_group` and `infinitesimal` work. Both of them only use one solution of `eq=0`.
https://github.com/sympy/sympy/blob/4c57a100981c066b7b15caf7808c7cdc6e737a51/sympy/solvers/ode.py#L6034
 https://github.com/sympy/sympy/blob/4c57a100981c066b7b15caf7808c7cdc6e737a51/sympy/solvers/ode.py#L5791

So due to this, only one solution is coming as an answer.

**Comment by RituRajSingh878** [comment]
```
diff --git a/sympy/solvers/ode.py b/sympy/solvers/ode.py
index 56b6acc..47e1625 100644
--- a/sympy/solvers/ode.py
+++ b/sympy/solvers/ode.py
@@ -5767,6 +5767,8 @@ def ode_lie_group(eq, func, order, match):
 
     """
 
+    if not match:
+        match = {'xi': None, 'eta': None}
     heuristics = lie_heuristics
     inf = {}
     f = func.func
@@ -5777,6 +5779,7 @@ def ode_lie_group(eq, func, order, match):
     xis = match.pop('xi')
     etas = match.pop('eta')
 
+
     if match:
         h = -simplify(match[match['d']]/match[match['e']])
         y = match['y']
@@ -5785,6 +5788,8 @@ def ode_lie_group(eq, func, order, match):
             sol = solve(eq, df)
             if sol == []:
                 raise NotImplementedError
+            elif len(sol)>1:
+                return [ode_lie_group(Eq(df, i), func, order, match) for i in sol]
         except NotImplementedError:
             raise NotImplementedError("Unable to solve the differential equation " +
                 str(eq) + " by the lie group method")
@@ -5835,6 +5840,8 @@ def ode_lie_group(eq, func, order, match):
                 scoord = newcoord[-1]
                 try:
                     sol = solve([r - rcoord, s - scoord], x, y, dict=True)
+                    if sol == []:
+                        raise NotImplementedError
                 except NotImplementedError:
                     continue
                 else:
```

I have modify the code of `lie_group` and it is working for the above ode if we will solve it by `lie_group` method

**Comment by RituRajSingh878** [comment]
```
In [1]: eq = (f(x).diff(x)-f(x)) * (f(x).diff(x)+f(x))                          

In [2]: dsolve(eq, hint='lie_group')                                            
Out[2]: 
⎡           -x             x⎤
⎣f(x) = C₁⋅ℯ  , f(x) = C₁⋅ℯ ⎦

In [3]: eq = f(x).diff(x) * (f(x).diff(x) - f(x))                               

In [4]: dsolve(eq, hint='lie_group')                                            
Out[4]: 
⎡                      x⎤
⎣f(x) = C₁, f(x) = C₁⋅ℯ ⎦

In [7]: eq = (f(x).diff(x)-f(x)) * (f(x).diff(x)+f(x))                          

In [8]: dsolve(eq, hint='lie_group')                                            
Out[8]: 
⎡           -x             x⎤
⎣f(x) = C₁⋅ℯ  , f(x) = C₁⋅ℯ ⎦

```

**Comment by RituRajSingh878** [comment]
> ```
> diff --git a/sympy/solvers/ode.py b/sympy/solvers/ode.py
> index 56b6acc..47e1625 100644
> --- a/sympy/solvers/ode.py
> +++ b/sympy/solvers/ode.py
> @@ -5767,6 +5767,8 @@ def ode_lie_group(eq, func, order, match):
>  
>      """
>  
> +    if not match:
> +        match = {'xi': None, 'eta': None}
>      heuristics = lie_heuristics
>      inf = {}
>      f = func.func
> @@ -5777,6 +5779,7 @@ def ode_lie_group(eq, func, order, match):
>      xis = match.pop('xi')
>      etas = match.pop('eta')
>  
> +
>      if match:
>          h = -simplify(match[match['d']]/match[match['e']])
>          y = match['y']
> @@ -5785,6 +5788,8 @@ def ode_lie_group(eq, func, order, match):
>              sol = solve(eq, df)
>              if sol == []:
>                  raise NotImplementedError
> +            elif len(sol)>1:
> +                return [ode_lie_group(Eq(df, i), func, order, match) for i in sol]
>          except NotImplementedError:
>              raise NotImplementedError("Unable to solve the differential equation " +
>                  str(eq) + " by the lie group method")
> @@ -5835,6 +5840,8 @@ def ode_lie_group(eq, func, order, match):
>                  scoord = newcoord[-1]
>                  try:
>                      sol = solve([r - rcoord, s - scoord], x, y, dict=True)
> +                    if sol == []:
> +                        raise NotImplementedError
>                  except NotImplementedError:
>                      continue
>                  else:
> ```
> 
> I have modified the code of `lie_group` and it is working for the above ode if we will solve it by `lie_group` method. If it is worth to add I will make a pr for this.

Or maybe we can do in a way that in dsolve we will factor the given equation and then pass one by one -

```
--- a/sympy/solvers/ode.py
+++ b/sympy/solvers/ode.py
@@ -632,6 +632,20 @@ def recur_len(l):
                 return [sol.subs(solved_constants) for sol in sols]
             return sols
     else:
+        from sympy.polys.polytools import factor
+        eqs = factor(eq)
+        if isinstance(eqs, Mul):
+            fac = eqs.args
+            eqs = []
+            for i in fac:
+                if i.has(Derivative):
+                    eqs.append(i)
+            if len(eqs) == 1:
+                return dsolve(eqs[0], func=None, hint="default", simplify=True,
+                       ics= None, xi=None, eta=None, x0=0, n=6, **kwargs)
+            else:
+                return [dsolve(eq, func=None, hint="default", simplify=True,
+                       ics= None, xi=None, eta=None, x0=0, n=6, **kwargs) for eq in eqs]
         given_hint = hint  # hint given by the user
 
         # See the docstring of _desolve for more details.
```

**Comment by oscarbenjamin** [comment]
I see that you have opened #17338 to solve this (it's good to comment on an issue about a PR that might solve it)

**Comment by oscarbenjamin** [comment]
@RituRajSingh878 is this issue fixed now that #17338 is merged?

The issue with integration constants remains but we have other open issues for that. Probably we should just use Dummy symbols for integration constants.

**Comment by oscarbenjamin** [comment]
Closing as fixed
